### ISSUE TITLE
bf16 is only supported on A100+ GPUs

### ISSUE THREAD

**Issue by snuffcn** [XLabs-AI/x-flux-comfyui#24]
Error occurred when executing XlabsSampler:

No operator found for `memory_efficient_attention_forward` with inputs:
query : shape=(24, 2560, 1, 128) (torch.bfloat16)
key : shape=(24, 2560, 1, 128) (torch.bfloat16)
value : shape=(24, 2560, 1, 128) (torch.bfloat16)
attn_bias :
p : 0.0
`decoderF` is not supported because:
attn_bias type is
bf16 is only supported on A100+ GPUs
`flshattF@v2.5.7` is not supported because:
requires device with capability > (8, 0) but your GPU has capability (7, 5) (too old)
bf16 is only supported on A100+ GPUs
`cutlassF` is not supported because:
bf16 is only supported on A100+ GPUs
`smallkF` is not supported because:
max(query.shape[-1] != value.shape[-1]) > 32
dtype=torch.bfloat16 (supported: {torch.float32})
bf16 is only supported on A100+ GPUs
unsupported embed per head: 128

**Comment by Vovanm88** [comment]
will check this

**Comment by snuffcn** [comment]
Thank you! My device is a 2080ti 22G and it's true that bf16 is not supported, I just don't understand what this XlabsSampler has to do with bf16.

**Comment by snuffcn** [comment]
Resolved
Line 288 of nodes.py

        device=mm.get_torch_device()
        dtype_model = torch.float16#dtype_model = torch.bfloat16#model.model.diffusion_model.img_in.weight.dtype

**Comment by Vovanm88** [comment]
fixed

**Comment by snuffcn** [comment]
device=mm.get_torch_device()
        if torch.cuda.is_bf16_supported():  <---------------
            dtype_model = torch.bfloat16#
        else:
            dtype_model = torch.float16#
Thank you~ it has been updated.
But this judgment is invalid, I had to manually change the code once again

**Comment by upboy-web** [comment]
Error occurred when executing XlabsSampler:

No operator found for `memory_efficient_attention_forward` with inputs:
query : shape=(24, 4352, 1, 128) (torch.bfloat16)
key : shape=(24, 4352, 1, 128) (torch.bfloat16)
value : shape=(24, 4352, 1, 128) (torch.bfloat16)
attn_bias :
p : 0.0
`decoderF` is not supported because:
attn_bias type is
bf16 is only supported on A100+ GPUs
`flshattF@v2.5.7` is not supported because:
requires device with capability > (8, 0) but your GPU has capability (7, 5) (too old)
bf16 is only supported on A100+ GPUs
`cutlassF-pt` is not supported because:
bf16 is only supported on A100+ GPUs
`smallkF` is not supported because:
max(query.shape[-1] != value.shape[-1]) > 32
dtype=torch.bfloat16 (supported: {torch.float32})
bf16 is only supported on A100+ GPUs
unsupported embed per head: 128

File "D:\AIGC\ComfyUI\execution.py", line 316, in execute
output_data, output_ui, has_subgraph = get_output_data(obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\execution.py", line 191, in get_output_data
return_values = _map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\execution.py", line 168, in _map_node_over_list
process_inputs(input_dict, i)
File "D:\AIGC\ComfyUI\execution.py", line 157, in process_inputs
results.append(getattr(obj, func)(**inputs))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\custom_nodes\x-flux-comfyui\nodes.py", line 353, in sampling
x = denoise_controlnet(
^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\custom_nodes\x-flux-comfyui\sampling.py", line 264, in denoise_controlnet
pred = model_forward(
^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\custom_nodes\x-flux-comfyui\sampling.py", line 41, in model_forward
img, txt = block(img=img, txt=txt, vec=vec, pe=pe)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\torch\nn\modules\module.py", line 1553, in _wrapped_call_impl
return self._call_impl(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\torch\nn\modules\module.py", line 1562, in _call_impl
return forward_call(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\comfy\ldm\flux\layers.py", line 166, in forward
attn = attention(torch.cat((txt_q, img_q), dim=2),
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\comfy\ldm\flux\math.py", line 11, in attention
x = optimized_attention(q, k, v, heads, skip_reshape=True)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\ComfyUI\comfy\ldm\modules\attention.py", line 380, in attention_xformers
out = xformers.ops.memory_efficient_attention(q, k, v, attn_bias=mask)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\xformers\ops\fmha\__init__.py", line 276, in memory_efficient_attention
return _memory_efficient_attention(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\xformers\ops\fmha\__init__.py", line 395, in _memory_efficient_attention
return _memory_efficient_attention_forward(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\xformers\ops\fmha\__init__.py", line 414, in _memory_efficient_attention_forward
op = _dispatch_fw(inp, False)
^^^^^^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\xformers\ops\fmha\dispatch.py", line 119, in _dispatch_fw
return _run_priority_list(
^^^^^^^^^^^^^^^^^^^
File "D:\AIGC\donghan\Lib\site-packages\xformers\ops\fmha\dispatch.py", line 55, in _run_priority_list
raise NotImplementedError(msg)

**Comment by zergqyue** [comment]
这个代码判断有问题 torch.cuda.is_bf16_supported()  ，我是2080ti 我的显卡不支持bf16，他也会返回true，所以把这个判断的代码删了就好了。  如果还报bf16相关的错误，就把bfloat16全部替换成float16就行了

**Comment by zergqyue** [comment]
我把xformers 方案换成sdp就不报错了

**Comment by lonrencn** [comment]
/ComfyUI/custom_nodes/x-flux-comfyui/nodes.py
line 301
        if torch.backends.mps.is_available():
            device = torch.device("mps")
        if torch.cuda.is_bf16_supported():
            dtype_model = torch.bfloat16#     <---------this question    edit --->dtype_model = torch.float16#
        else:
            dtype_model = torch.float16#



result:
        if torch.backends.mps.is_available():
            device = torch.device("mps")
        if torch.cuda.is_bf16_supported():
            dtype_model = torch.float16#
        else:
            dtype_model = torch.float16#

reboot comfyui ok!!~

**Comment by amir8alizadeh** [comment]
> 我把xformers 方案换成sdp就不报错了

Please explain more

**Comment by wsj998689** [comment]
1、没有使用gpu，检查一下CUDA_VISIBLE_DEVICES=0是否写正确
2、安装了flash-attn，卸载试一试
